### DISCUSSION TITLE
a few extra optimizations i'd like to suggest

### DISCUSSION THREAD

**Discussion by eadmaster** [MichaIng/DietPi#5934] [Q&A] [answered]
````
# avoid installing unrequired langpacks
apt install localepurge

# autoclean apt deb cache https://askubuntu.com/questions/389728/how-can-i-configure-apt-get-to-clean-automatically-after-every-install
echo "APT::Keep-Downloaded-Packages "false";" > /etc/apt/apt.conf.d/99-autoclean
echo "Binary::apt::APT::keep-downloaded-packages "false";" >> /etc/apt/apt.conf.d/99-autoclean	
````

Further optimizations to reduce disk usage:

1. disable the journal on ext4: `tune2fs -O ^has_journal  /dev/mmcblk0p1`, or
2. switch to btrfs with zstd compression. I think it easier @ [build time via `ROOTFS_TYPE=btrfs; BTRFS_COMPRESSION=zstd`](https://docs.armbian.com/Developer-Guide_Build-Options/)

**Comment by MichaIng** [comment, marked as answer]
> avoid installing unrequired langpacks

There are no locales installed on DietPi images, aside of the mandatory `C`, `C.UTF-8` and `POSIX`. And when you select a language via `dietpi-config`, only the single additional locale is compiled while others are removed. So installing `localepuge` only adds additional data (the package) while it does not remove anything.

> autoclean apt deb cache

This is the default already:
```console
# apt-config dump | grep -i downloaded
Binary::apt::APT::Keep-Downloaded-Packages "0";
```

> disable the journal on ext4

The ext4 journal is THE huge advantage of ext4 vs older variants, lowering the risk of permanent data loss dramatically. ~75% of our users still use SD cards, so any help against filesystem corruption and data loss is important. So this will definitely stay enabled on images we ship.

> switch to btrfs with zstd compression

This lowers I/O performance and raises CPU usage, not feasible in many use cases and low power SBCs.

> [ROOTFS_TYPE=btrfs; BTRFS_COMPRESSION=zstd](https://docs.armbian.com/Developer-Guide_Build-Options/)

We do not use the Armbian build scripts, but our own. Enabling compression on Btrfs is however trivial, I'm just not sure if this is really a good default.

But we may provide F2FS and Btrfs rootfs images in the future. F2FS on RPi is simple, since it is compiled into the kernel, but Btrfs is not. Hence for RPi it would require adding an initramfs, which again raises disk usage and I/O (on kernel upgrades). On other SBCs, an initramfs is used already, but a single partition only. For a F2FS and Btrfs rootfs partition, it requires a dedicated boot partition where kernel, dtb and initramfs are located.

**Comment by eadmaster** [comment]
> There are no locales installed on DietPi images, aside of the mandatory C, C.UTF-8 and POSIX. And when you select a language via dietpi-config, only the single additional locale is compiled while others are removed. S

The purpose of localepurge is avoid installing the extra locales that are bundled with deb packages. They get installed regardless of actual system locale, just check `/usr/share/locale`.

Another path that may use some cleanup is `/usr/share/doc`, usually filled with unnecessary licenses, changelogs, etc.

> The ext4 journal is THE huge advantage of ext4 vs older variants, lowering the risk of permanent data loss dramatically. ~75% of our users still use SD cards, so any help against filesystem corruption and data loss is important. So this will definitely stay enabled on images we ship.

ok, i agree with having the safer default, it can still be easily disabled without reformatting.

> But we may provide F2FS and Btrfs rootfs images in the future. F2FS on RPi is simple, since it is compiled into the kernel, but Btrfs is not.

Btrfs is listed in  `/proc/filesystems` in the Bullseye image. I don't know about the boot process details, but i guess F2FS+zstd would be a good alternative, so i'm definitively interested in this.

**Reply by MichaIng** [reply]
> The purpose of localepurge is avoid installing the extra locales that are bundled with deb packages.

Ah, those you mean. We cannot ship our images with those remove, of course and we do not know whether users might switch back and forth the language, in which case all locales are lost until all related packages have been upgraded or reinstalled. Also, as the package description states, it can cause issues in APT/dpkg, maintainer scripts etc, so should be installed/used with case. And:

> It is run automatically upon completion of any APT installation actions.

It seems to be done inefficiently. I have those files removed on all my production systems as well, but I use:
```console
root@dietpi:~# grep locale /etc/dpkg/dpkg.cfg.d/01-micha
# Skip locales all together
path-exclude /usr/share/locale/*
path-exclude /etc/locale.alias
```
That way those files are not installed in the first place, so they do not need to be removed afterwards, and nothing needs to run. Of course the files need to be removed manually once.

All those looks more like a feature for `dietpi-cleaner`, where we can warn about potential issues. Leaving individual language files in place, can be offered or done based on current locale via:
```
path-exclude /usr/share/locale/*
path-include /usr/share/locale/de*
```

> Another path that may use some cleanup is `/usr/share/doc`, usually filled with unnecessary licenses, changelogs, etc.

The license files must be kept, at least by everything we implement, for legal reasons. Also this can contain example/default configs and such, so is often of more interest. But otherwise similar: An offer to have those removed and excluded from dpkg could be implemented into `dietpi-cleaner`, and the same for a bunch of other optional files, e.g. on some of my production systems:
```
path-exclude /usr/share/doc/*
# We need to keep copyright files for legal reasons
path-include /usr/share/doc/*/copyright
path-exclude /usr/share/doc-base/*
path-exclude /usr/share/doc-base
path-exclude /usr/share/groff/*
path-exclude /usr/share/groff
path-exclude /usr/share/info/*
path-exclude /usr/share/info
path-exclude /usr/share/linda/*
path-exclude /usr/share/linda
path-exclude /usr/share/lintian/*
path-exclude /usr/share/lintian
path-exclude /usr/share/man/*
path-exclude /usr/share/man
path-exclude /etc/alternatives/*.gz
path-exclude /var/lib/dpkg/alternatives/*.gz
```

> Btrfs is listed in `/proc/filesystems` in the Bullseye image. I don't know about the boot process details, but i guess F2FS+zstd would be a good alternative, so i'm definitively interested in this.

In my case it's not, you can check whether it's a dedicated module via `modinfo`:
```console
root@micha:~# grep btrfs /proc/filesystems
root@micha:~# modinfo btrfs
filename:       /lib/modules/5.15.76-v7+/kernel/fs/btrfs/btrfs.ko.xz
softdep:        pre: blake2b-256
softdep:        pre: sha256
softdep:        pre: xxhash64
softdep:        pre: crc32c
license:        GPL
alias:          devname:btrfs-control
alias:          char-major-10-234
alias:          fs-btrfs
srcversion:     FE0637D96B52A3D888CC47F
depends:        zstd_compress,raid6_pq,xor
intree:         Y
name:           btrfs
vermagic:       5.15.76-v7+ SMP mod_unload modversions ARMv7 p2v8
```
Since `/lib/modules` would be inside the Btrfs filesystem, you have a bootstrap cycle. This is what an initramfs is for: It is compiled with the required filesystem module(s) included and stored on the boot partition, and can then mount the root partition for the kernel.

> but i guess F2FS+zstd would be a good alternative

Jep, pretty similar to Btrfs, simple to enable. `dietpi-drive_manager` needs to be adjusted to preserve those mount options, respectively allow to choose those explicitly.

**Comment by eadmaster** [comment]
> It seems to be done inefficiently. I have those files removed on all my production systems as well, but I use:

[localepurge is also using dpkg filters by default](https://sleeplessbeastie.eu/2018/09/03/how-to-remove-useless-localizations/#additional-notes), and provides a gui for the user to select which locales to keep.

> In my case it's not, you can check whether it's a dedicated module via modinfo

In the [NanoPiR1 Bullseye image](https://dietpi.com/downloads/images/DietPi_NanoPiR1-ARMv7-Bullseye.7z) is builtin for some reason:
````
# modinfo btrfs
name:           btrfs
filename:       (builtin)
softdep:        pre: blake2b-256
softdep:        pre: sha256
softdep:        pre: xxhash64
softdep:        pre: crc32c
license:        GPL
file:           fs/btrfs/btrfs
alias:          devname:btrfs-control
alias:          char-major-10-234
alias:          fs-btrfs
````

**Reply by MichaIng** [reply]
> localepurge is also using dpkg filters by default

Ah nice. I'd probably still ship an own implementation, just because I'm doing this stuff on my own systems already, and would then implement the same for documentational files in the same turn. But great that there is a proper solution already.

> In the [NanoPiR1 Bullseye image](https://dietpi.com/downloads/images/DietPi_NanoPiR1-ARMv7-Bullseye.7z) is builtin for some reason:

Ah, probably it is with Armbian kernels, I've checked on RPi. F2FS is not builtin, is it?
```sh
modinfo f2fs
```
The question then is whether U-Boot itself can read kernel etc from Btrfs or F2FS as well, or whether a dedicated boot partition with FAT or ext4 filesystem is still required. Can be tested when I find time to do last adjustments to our build script to assure both filesystems are properly supported.

So to sum up:
- Images with F2FS and/or Btrfs, basically this issue generalised: #606
- Support for native filesystem compression, basically this issue generalised: #3863
- Support for removal and dpkg-exclusion of non-essential system files (via `dietpi-cleaner`, I'd say) => [these directories](https://github.com/MichaIng/DietPi/discussions/5934?notification_referrer_id=NT_kwDOAbKUwbM0OTU5MTQ2MzkxOjI4NDgwNzA1#discussioncomment-4295867)

Disabling ext4 journal is not something I'd ever recommend or provide a feature for, or only in form of a generic option to toggle all ext4 filesystem feature in general.

Did you observe a case where APT packages were not removed from cache automatically after successful install, without setting `APT::Keep-Downloaded-Packages` explicitly?

**Comment by eadmaster** [comment]
> Ah, probably it is with Armbian kernels, I've checked on RPi. F2FS is not builtin, is it?

It is:
````
# modinfo f2fs
name:           f2fs
filename:       (builtin)
softdep:        pre: crc32
license:        GPL
file:           fs/f2fs/f2fs
description:    Flash Friendly File System
author:         Samsung Electronics's Praesto Team
alias:          fs-f2fs
parm:           num_compress_pages:Number of intermediate compress pages to preallocate (uint)

````

> Support for removal and dpkg-exclusion of non-essential system files (via dietpi-cleaner, I'd say) => https://github.com/MichaIng/DietPi/discussions/5934?notification_referrer_id=NT_kwDOAbKUwbM0OTU5MTQ2MzkxOjI4NDgwNzA1#discussioncomment-4295867

Why not simply adding an option to run `apt install localepurge` from `dietpi-cleaner`?

> Did you observe a case where APT packages were not removed from cache automatically after successful install, without setting APT::Keep-Downloaded-Packages explicitly?

No, but they get cleaned only when the install process is completed, which confused me initially.

**Reply by MichaIng** [reply]
> It is:

Ah great, this makes it easier.

> Why not simply adding an option to run `apt install localepurge` from `dietpi-cleaner`?

Because the cleaner shall cleanup and not install additional packages 😄. But mostly, because I'll implement the other info/man dirs the same way then, so code/methods can be shared, and everything is under our control, not relying on 3rd party software development, and additional dependencies like `perl`.

> No, but they get cleaned only when the install process is completed, which confused me initially.

Ah yeah, if the install is aborted, the downloaded packages remain. Although this might be the wanted behaviour, as a second install attempt won't need to download packages again.
